// @cloudinary-util/url-loader/dist/lib/cloudinary.js
// src/lib/cloudinary.ts
import { objectHasKey as objectHasKey3, parseUrl } from "@cloudinary-util/util";
import {
  Cloudinary
} from "@cloudinary/url-gen";

// src/lib/plugin.ts
var plugin = (def) => ({
  strict: false,
  alwaysApply: false,
  ...def,
  apply: (cldAsset, ctx) => def.apply(cldAsset, ctx, ctx)
});

// src/plugins/abr.ts
var AbrPlugin = /* @__PURE__ */ plugin({
  name: "Abr",
  supports: "video",
  inferOwnOptions: {},
  props: {
    streamingProfile: true
  },
  apply: (asset, opts) => {
    if (typeof opts.streamingProfile !== "string") return {};
    asset.addTransformation(`sp_${opts.streamingProfile}`);
    return {};
  }
});

// src/plugins/default-image.ts
import { getFormat } from "@cloudinary-util/util";
var DefaultImagePlugin = /* @__PURE__ */ plugin({
  name: "DefaultImage",
  supports: "image",
  inferOwnOptions: {},
  props: { defaultImage: true },
  apply: (asset, opts) => {
    const { defaultImage } = opts;
    if (typeof defaultImage !== "string") return {};
    if (!getFormat(defaultImage)) {
      console.warn(
        `The defaultImage prop may be missing a format and must include it along with the public ID. (Ex: myimage.jpg)`
      );
    }
    const defaultImageId = defaultImage.replace(/\//g, ":");
    asset.addTransformation(`d_${defaultImageId}`);
    return {};
  }
});

// src/constants/qualifiers.ts
import { convertColorHexToRgb, testColorIsHex } from "@cloudinary-util/util";
var convertersColors = [
  {
    test: testColorIsHex,
    convert: convertColorHexToRgb
  }
];
var primary = {
  aspectRatio: {
    qualifier: "ar"
  },
  crop: {
    qualifier: "c"
  },
  gravity: {
    qualifier: "g"
  },
  height: {
    qualifier: "h"
  },
  width: { qualifier: "w" }
};
var position = {
  angle: {
    qualifier: "a"
  },
  gravity: {
    qualifier: "g"
  },
  x: { qualifier: "x" },
  y: { qualifier: "y" }
};
var text = {
  alignment: {
    qualifier: false,
    order: 6
  },
  antialias: {
    qualifier: "antialias"
  },
  border: {
    qualifier: "bo",
    location: "primary"
  },
  color: {
    qualifier: "co",
    location: "primary",
    converters: convertersColors
  },
  fontFamily: {
    qualifier: false,
    order: 1
  },
  fontSize: {
    qualifier: false,
    order: 2
  },
  fontStyle: {
    qualifier: false,
    order: 4
  },
  fontWeight: {
    qualifier: false,
    order: 3
  },
  hinting: {
    qualifier: "hinting"
  },
  letterSpacing: {
    qualifier: "letter_spacing"
  },
  lineSpacing: {
    qualifier: "line_spacing"
  },
  stroke: {
    qualifier: "self",
    order: 7
  },
  textDecoration: {
    qualifier: false,
    order: 5
  }
};
var effects = {
  angle: {
    qualifier: "a"
  },
  art: {
    prefix: "e",
    qualifier: "art"
  },
  autoBrightness: {
    prefix: "e",
    qualifier: "auto_brightness"
  },
  autoColor: {
    prefix: "e",
    qualifier: "auto_color"
  },
  autoContrast: {
    prefix: "e",
    qualifier: "auto_contrast"
  },
  assistColorblind: {
    prefix: "e",
    qualifier: "assist_colorblind"
  },
  background: {
    qualifier: "b"
  },
  blackwhite: {
    prefix: "e",
    qualifier: "blackwhite"
  },
  blur: {
    prefix: "e",
    qualifier: "blur"
  },
  blurFaces: {
    prefix: "e",
    qualifier: "blur_faces"
  },
  blurRegion: {
    prefix: "e",
    qualifier: "blur_region"
  },
  border: {
    qualifier: "bo"
  },
  brightness: {
    prefix: "e",
    qualifier: "brightness"
  },
  brightnessHSB: {
    prefix: "e",
    qualifier: "brightness_hsb"
  },
  cartoonify: {
    prefix: "e",
    qualifier: "cartoonify"
  },
  color: {
    qualifier: "co",
    converters: convertersColors
  },
  colorize: {
    prefix: "e",
    qualifier: "colorize"
  },
  contrast: {
    prefix: "e",
    qualifier: "contrast"
  },
  displace: {
    prefix: "e",
    qualifier: "distort"
  },
  distort: {
    prefix: "e",
    qualifier: "distort"
  },
  fillLight: {
    prefix: "e",
    qualifier: "fill_light"
  },
  gamma: {
    prefix: "e",
    qualifier: "gamma"
  },
  gradientFade: {
    prefix: "e",
    qualifier: "gradient_fade"
  },
  grayscale: {
    prefix: "e",
    qualifier: "grayscale"
  },
  hue: {
    prefix: "e",
    qualifier: "hue"
  },
  improve: {
    prefix: "e",
    qualifier: "improve"
  },
  loop: {
    prefix: "e",
    qualifier: "loop"
  },
  multiply: {
    prefix: "e",
    qualifier: "multiply"
  },
  negate: {
    prefix: "e",
    qualifier: "negate"
  },
  noise: {
    prefix: "e",
    qualifier: "noise"
  },
  oilPaint: {
    prefix: "e",
    qualifier: "oil_paint"
  },
  opacity: {
    qualifier: "o"
  },
  outline: {
    prefix: "e",
    qualifier: "outline"
  },
  pixelate: {
    prefix: "e",
    qualifier: "pixelate"
  },
  pixelateFaces: {
    prefix: "e",
    qualifier: "pixelate_faces"
  },
  pixelateRegion: {
    prefix: "e",
    qualifier: "pixelate_region"
  },
  radius: {
    qualifier: "r"
  },
  redeye: {
    prefix: "e",
    qualifier: "redeye"
  },
  replaceColor: {
    prefix: "e",
    qualifier: "replace_color"
  },
  saturation: {
    prefix: "e",
    qualifier: "saturation"
  },
  screen: {
    prefix: "e",
    qualifier: "screen"
  },
  sepia: {
    prefix: "e",
    qualifier: "sepia"
  },
  shadow: {
    prefix: "e",
    qualifier: "shadow"
  },
  sharpen: {
    prefix: "e",
    qualifier: "sharpen"
  },
  shear: {
    prefix: "e",
    qualifier: "shear"
  },
  simulateColorblind: {
    prefix: "e",
    qualifier: "simulate_colorblind"
  },
  tint: {
    prefix: "e",
    qualifier: "tint"
  },
  trim: {
    prefix: "e",
    qualifier: "trim"
  },
  unsharpMask: {
    prefix: "e",
    qualifier: "unsharp_mask"
  },
  vectorize: {
    prefix: "e",
    qualifier: "vectorize"
  },
  vibrance: {
    prefix: "e",
    qualifier: "vibrance"
  },
  vignette: {
    prefix: "e",
    qualifier: "vignette"
  }
};

// src/lib/transformations.ts
function constructTransformation({
  prefix,
  qualifier,
  value,
  converters
}) {
  let transformation = "";
  if (prefix) {
    transformation = `${prefix}_`;
  }
  let transformationValue = value;
  converters?.forEach(({ test, convert }) => {
    if (!test(transformationValue)) return;
    transformationValue = convert(transformationValue);
  });
  if (transformationValue === true || transformationValue === "true") {
    return `${transformation}${qualifier}`;
  }
  if (typeof transformationValue === "string" || typeof transformationValue === "number") {
    if (prefix) {
      return `${transformation}${qualifier}:${transformationValue}`;
    } else {
      return `${qualifier}_${transformationValue}`;
    }
  }
}
function promptArrayToString(promptArray) {
  return `(${promptArray.join(";")})`;
}

// src/lib/utils.ts
var isArray = Array.isArray;
var entriesOf = Object.entries;
var throwError = (message, ctor = Error) => {
  throw new ctor(message);
};

// src/plugins/effects.ts
var EffectsPlugin = /* @__PURE__ */ plugin({
  name: "Effects",
  supports: "all",
  inferOwnOptions: {},
  props: {
    angle: true,
    art: true,
    autoBrightness: true,
    autoColor: true,
    autoContrast: true,
    assistColorblind: true,
    background: true,
    blackwhite: true,
    blur: true,
    blurFaces: true,
    blurRegion: true,
    border: true,
    brightness: true,
    brightnessHSB: true,
    cartoonify: true,
    color: true,
    colorize: true,
    contrast: true,
    displace: true,
    distort: true,
    fillLight: true,
    gamma: true,
    gradientFade: true,
    grayscale: true,
    hue: true,
    improve: true,
    loop: true,
    multiply: true,
    negate: true,
    noise: true,
    oilPaint: true,
    opacity: true,
    outline: true,
    pixelate: true,
    pixelateFaces: true,
    pixelateRegion: true,
    radius: true,
    redeye: true,
    replaceColor: true,
    saturation: true,
    screen: true,
    sepia: true,
    shadow: true,
    sharpen: true,
    shear: true,
    simulateColorblind: true,
    tint: true,
    trim: true,
    unsharpMask: true,
    vectorize: true,
    vibrance: true,
    vignette: true,
    effects: true
  },
  apply: (cldAsset, opts) => {
    const transformationStrings = constructTransformationString({
      effects,
      options: opts
    });
    transformationStrings.forEach((transformation) => {
      if (transformation) {
        cldAsset.addTransformation(transformation);
      }
    });
    if (isArray(opts?.effects)) {
      opts?.effects.forEach((effectsSet) => {
        const transformationString = constructTransformationString({
          effects,
          options: effectsSet
        }).filter((t) => !!t).join(",");
        cldAsset.addTransformation(transformationString);
      });
    }
    function constructTransformationString({
      effects: effects2,
      options
    }) {
      return Object.keys(effects2).map(
        (key) => {
          const { prefix, qualifier, converters } = effects2[key];
          return constructTransformation({
            qualifier,
            prefix,
            value: options?.[key],
            converters
          });
        }
      );
    }
    return {};
  }
});

// src/plugins/fill-background.ts
import { normalizeNumberParameter } from "@cloudinary-util/util";
var defaultCrop = "pad";
var FillBackgroundPlugin = /* @__PURE__ */ plugin({
  name: "FillBackground",
  supports: "image",
  inferOwnOptions: {},
  props: {
    fillBackground: true
  },
  apply: (cldAsset, opts, ctx) => {
    const { fillBackground } = opts;
    if (typeof fillBackground === "undefined") return {};
    const width = normalizeNumberParameter(ctx.width);
    const height = normalizeNumberParameter(ctx.height);
    const hasDefinedDimensions = typeof height === "number" && typeof width === "number";
    let aspectRatio = ctx.aspectRatio;
    if (!aspectRatio && hasDefinedDimensions) {
      aspectRatio = `${width}:${height}`;
    }
    if (!aspectRatio) {
      if (process.env.NODE_ENV === "development") {
        console.warn(
          `Could not determine aspect ratio based on available options to use fillBackground. Please specify width and height or an aspect ratio.`
        );
      }
      return {};
    }
    if (fillBackground === true) {
      const properties = [
        "b_gen_fill",
        `ar_${aspectRatio}`,
        `c_${defaultCrop}`
      ];
      cldAsset.addTransformation(properties.join(","));
    } else if (typeof fillBackground === "object") {
      const { crop = defaultCrop, gravity, prompt } = fillBackground;
      const properties = [`ar_${aspectRatio}`, `c_${crop}`];
      if (typeof prompt === "string") {
        properties.unshift(`b_gen_fill:${prompt}`);
      } else {
        properties.unshift(`b_gen_fill`);
      }
      if (typeof gravity === "string") {
        properties.push(`g_${gravity}`);
      }
      cldAsset.addTransformation(properties.join(","));
    }
    return {};
  }
});

// src/plugins/flags.ts
var FlagsPlugin = /* @__PURE__ */ plugin({
  name: "Flags",
  supports: "all",
  inferOwnOptions: {},
  props: {
    flags: true
  },
  apply: (cldAsset, { flags }) => {
    if (typeof flags === "string") {
      flags = [flags];
    }
    if (isArray(flags)) {
      flags.forEach((flag) => cldAsset.addFlag(flag));
    } else if (typeof flags === "object") {
      Object.entries(flags).forEach(([qualifier, value]) => {
        cldAsset.addTransformation(`fl_${qualifier}:${value}`);
      });
    }
    return {};
  }
});

// src/plugins/preserve-transformations.ts
import { getTransformations } from "@cloudinary-util/util";
var PreserveTransformationsPlugin = /* @__PURE__ */ plugin({
  name: "PreserveTransformations",
  supports: "all",
  inferOwnOptions: {},
  props: {
    preserveTransformations: true
  },
  apply: (cldAsset, opts, ctx) => {
    const { preserveTransformations = false } = opts;
    if (preserveTransformations) {
      try {
        if (ctx.src === void 0) {
          throw new Error("options.src was undefined");
        }
        const transformations = getTransformations(ctx.src).map(
          (t) => t.join(",")
        );
        transformations.flat().forEach((transformation) => {
          cldAsset.addTransformation(transformation);
        });
      } catch (e) {
        console.warn(
          `Failed to preserve transformations: ${e.message}`
        );
      }
    }
    return {};
  }
});

// src/plugins/raw-transformations.ts
var RawTransformationsPlugin = /* @__PURE__ */ plugin({
  name: "RawTransformations",
  supports: "all",
  inferOwnOptions: {},
  props: {
    rawTransformations: true
  },
  apply: (cldAsset, opts) => {
    let { rawTransformations = [] } = opts;
    if (!isArray(rawTransformations)) {
      rawTransformations = [rawTransformations];
    }
    rawTransformations.forEach((transformation) => {
      cldAsset.addTransformation(transformation);
    });
    return {};
  }
});

// src/plugins/recolor.ts
var RecolorPlugin = /* @__PURE__ */ plugin({
  name: "Recolor",
  supports: "image",
  inferOwnOptions: {},
  props: {
    recolor: true
  },
  apply: (cldAsset, opts) => {
    const { recolor } = opts;
    const recolorOptions = {
      prompt: void 0,
      "to-color": void 0,
      multiple: void 0
    };
    if (isArray(recolor)) {
      if (isArray(recolor[0])) {
        recolorOptions.prompt = promptArrayToString(recolor[0]);
      } else {
        recolorOptions.prompt = recolor[0];
      }
      if (typeof recolor[1] === "string") {
        recolorOptions["to-color"] = recolor[1];
      }
    } else if (typeof recolor === "object") {
      if (typeof recolor.prompt === "string") {
        recolorOptions.prompt = recolor.prompt;
      } else if (isArray(recolor.prompt)) {
        recolorOptions.prompt = promptArrayToString(recolor.prompt);
      }
      if (typeof recolor.to === "string") {
        recolorOptions["to-color"] = recolor.to;
      }
      if (recolor.multiple === true) {
        recolorOptions.multiple = `true`;
      }
    }
    const transformation = Object.entries(recolorOptions).filter(([, value]) => !!value).map(([key, value]) => `${key}_${value}`).join(";");
    if (transformation) {
      cldAsset.addTransformation(`e_gen_recolor:${transformation}`);
    }
    return {};
  }
});

// src/plugins/remove-background.ts
var RemoveBackgroundPlugin = /* @__PURE__ */ plugin({
  name: "RemoveBackground",
  supports: "image",
  inferOwnOptions: {},
  props: {
    removeBackground: true
  },
  apply: (cldAsset, opts) => {
    if (opts.removeBackground) {
      cldAsset.addTransformation("e_background_removal");
    }
    return {};
  }
});

// src/plugins/remove.ts
var RemovePlugin = /* @__PURE__ */ plugin({
  name: "Remove",
  supports: "image",
  inferOwnOptions: {},
  props: {
    remove: true
  },
  apply: (cldAsset, opts) => {
    const { remove } = opts;
    const removeOptions = {
      prompt: void 0,
      region: void 0,
      multiple: void 0,
      "remove-shadow": void 0
    };
    if (typeof remove === "string") {
      removeOptions.prompt = remove;
    } else if (isArray(remove)) {
      removeOptions.prompt = promptArrayToString(remove);
    } else if (typeof remove === "object") {
      const hasPrompt = typeof remove.prompt === "string" || isArray(remove.prompt);
      const hasRegion = isArray(remove.region);
      if (hasPrompt && hasRegion) {
        throw new Error(
          "Invalid remove options: you can not have both a prompt and a region. More info: https://cloudinary.com/documentation/transformation_reference#e_gen_remove"
        );
      }
      if (typeof remove.prompt === "string") {
        removeOptions.prompt = remove.prompt;
      } else if (isArray(remove.prompt)) {
        removeOptions.prompt = promptArrayToString(remove.prompt);
      }
      if (isArray(remove.region)) {
        removeOptions.region = regionArrayToString(remove.region);
      }
      if (remove.multiple === true) {
        removeOptions.multiple = `true`;
      }
      if (remove.removeShadow === true) {
        removeOptions["remove-shadow"] = `true`;
      }
    }
    const transformation = Object.entries(removeOptions).filter(([, value]) => !!value).map(([key, value]) => `${key}_${value}`).join(";");
    if (transformation) {
      cldAsset.addTransformation(`e_gen_remove:${transformation}`);
    }
    return {};
  }
});
function regionArrayToString(regionArray) {
  const indexes = {
    0: "x",
    1: "y",
    2: "w",
    3: "h"
  };
  const regionString = regionArray.map((region, index) => {
    if (isArray(region)) {
      return regionArrayToString(region);
    }
    const key = indexes[index];
    return `${key}_${region}`;
  }).join(";");
  return `(${regionString})`;
}

// src/plugins/replace-background.ts
var ReplaceBackgroundPlugin = /* @__PURE__ */ plugin({
  name: "ReplaceBackground",
  supports: "image",
  inferOwnOptions: {},
  props: {
    replaceBackground: true
  },
  apply: (cldAsset, opts) => {
    const { replaceBackground } = opts;
    if (!replaceBackground) return {};
    const properties = [];
    if (typeof replaceBackground === "object") {
      if (typeof replaceBackground.prompt !== "undefined") {
        properties.push(`prompt_${replaceBackground.prompt}`);
      }
      if (typeof replaceBackground.seed === "number") {
        properties.push(`seed_${replaceBackground.seed}`);
      }
    } else if (typeof replaceBackground === "string") {
      properties.push(`prompt_${replaceBackground}`);
    }
    let transformation = "e_gen_background_replace";
    if (properties.length > 0) {
      transformation = `${transformation}:${properties.join(";")}`;
    }
    cldAsset.addTransformation(transformation);
    return {};
  }
});

// src/plugins/replace.ts
var ReplacePlugin = /* @__PURE__ */ plugin({
  name: "Replace",
  supports: "image",
  inferOwnOptions: {},
  props: {
    replace: true
  },
  apply: (cldAsset, opts) => {
    const { replace } = opts;
    if (!replace) return {};
    let from, to, preserveGeometry = false;
    if (isArray(replace)) {
      from = replace[0];
      to = replace[1];
      preserveGeometry = replace[2] || false;
    } else {
      from = replace.from;
      to = replace.to;
      preserveGeometry = replace.preserveGeometry || false;
    }
    const properties = [`e_gen_replace:from_${from}`, `to_${to}`];
    if (preserveGeometry) {
      properties.push(`preserve-geometry_${preserveGeometry}`);
    }
    cldAsset.addTransformation(properties.join(";"));
    return {};
  }
});

// src/plugins/restore.ts
var RestorePlugin = /* @__PURE__ */ plugin({
  name: "Restore",
  supports: "image",
  inferOwnOptions: {},
  props: {
    restore: true
  },
  apply: (cldAsset, opts) => {
    const { restore } = opts;
    if (restore) {
      cldAsset.addTransformation("e_gen_restore");
    }
    return {};
  }
});

// src/plugins/sanitize.ts
var SanitizePlugin = /* @__PURE__ */ plugin({
  name: "Sanitize",
  supports: "image",
  inferOwnOptions: {},
  props: {
    sanitize: true
  },
  alwaysApply: true,
  apply: (cldAsset, opts, ctx) => {
    const { sanitize = true } = opts;
    const shouldApplySanitizer = sanitize && (ctx.format === "svg" || cldAsset.publicID.endsWith(".svg"));
    if (shouldApplySanitizer) {
      cldAsset.addTransformation("fl_sanitize");
    }
    return {};
  }
});

// src/plugins/seo.ts
var SeoPlugin = /* @__PURE__ */ plugin({
  name: "Seo",
  supports: "all",
  inferOwnOptions: {},
  props: {
    seoSuffix: true
  },
  apply: (cldAsset, opts, ctx) => {
    const { seoSuffix } = opts;
    if (typeof seoSuffix !== "string") return {};
    if (ctx.deliveryType === "fetch") {
      console.warn("SEO suffix is not supported with a delivery type of fetch");
    } else {
      cldAsset.setSuffix(seoSuffix);
    }
    return {};
  }
});

// src/plugins/underlays.ts
import { objectHasKey } from "@cloudinary-util/util";
var UnderlaysPlugin = /* @__PURE__ */ plugin({
  name: "Underlays",
  supports: "all",
  inferOwnOptions: {},
  props: {
    underlay: true,
    underlays: true
  },
  apply: (cldAsset, opts) => {
    const { underlay, underlays = [] } = opts;
    const typeQualifier = "u";
    if (isArray(underlays)) {
      underlays.forEach(applyUnderlay);
    }
    if (typeof underlay === "string") {
      const underlayOptions = {
        publicId: underlay,
        crop: "fill",
        width: "1.0",
        height: "1.0",
        flags: ["relative"]
      };
      applyUnderlay(underlayOptions);
    }
    function applyUnderlay({
      publicId,
      type,
      position: position2,
      effects: layerEffects = [],
      flags: layerFlags = [],
      appliedFlags = [],
      ...options
    }) {
      const hasPublicId = typeof publicId === "string";
      const hasPosition = typeof position2 === "object";
      if (!hasPublicId) {
        console.warn(`An ${type} is missing a Public ID`);
        return;
      }
      let layerTransformation = `${typeQualifier}_${publicId.replace(
        /\//g,
        ":"
      )}`;
      const primary2 = [];
      const applied = [];
      entriesOf(options).forEach(([key, value]) => {
        if (!objectHasKey(primary, key)) return;
        const { qualifier } = primary[key];
        primary2.push(`${qualifier}_${value}`);
      });
      layerEffects.forEach((effect) => {
        Object.keys(effect).forEach((key) => {
          if (!objectHasKey(primary, key)) return;
          const { qualifier } = primary[key];
          primary2.push(`${qualifier}_${effect[key]}`);
        });
      });
      if (hasPosition) {
        entriesOf(position2).forEach(([key, value]) => {
          if (!objectHasKey(position, key)) return;
          const { qualifier } = position[key];
          applied.push(`${qualifier}_${value}`);
        });
      }
      const activeLayerFlags = isArray(layerFlags) ? layerFlags : [layerFlags];
      activeLayerFlags.forEach((flag) => primary2.push(`fl_${flag}`));
      const activeAppliedFlags = isArray(appliedFlags) ? appliedFlags : [appliedFlags];
      activeAppliedFlags.forEach((flag) => applied.push(`fl_${flag}`));
      layerTransformation = `${layerTransformation},${primary2.join(",")}`;
      layerTransformation = `${layerTransformation}/fl_layer_apply,fl_no_overflow`;
      if (applied.length > 0) {
        layerTransformation = `${layerTransformation},${applied.join(",")}`;
      }
      cldAsset.addTransformation(layerTransformation);
    }
    return {};
  }
});

// src/plugins/zoompan.ts
var ZoompanPlugin = /* @__PURE__ */ plugin({
  name: "Zoompan",
  supports: "image",
  inferOwnOptions: {},
  props: {
    zoompan: true
  },
  apply: (cldAsset, opts) => {
    const { zoompan = false } = opts;
    const overrides = {
      format: void 0
    };
    if (zoompan === true) {
      cldAsset.addTransformation("e_zoompan");
    } else if (typeof zoompan === "string") {
      if (zoompan === "loop") {
        cldAsset.addTransformation("e_zoompan");
        cldAsset.addTransformation("e_loop");
      } else {
        cldAsset.addTransformation(`e_zoompan:${zoompan}`);
      }
    } else if (typeof zoompan === "object") {
      let zoompanEffect = "e_zoompan";
      if (typeof zoompan.options === "string") {
        zoompanEffect = `${zoompanEffect}:${zoompan.options}`;
      }
      cldAsset.addTransformation(zoompanEffect);
      let loopEffect;
      if (zoompan.loop === true) {
        loopEffect = "e_loop";
      } else if (typeof zoompan.loop === "string" || typeof zoompan.loop === "number") {
        loopEffect = `e_loop:${zoompan.loop}`;
      }
      if (loopEffect) {
        cldAsset.addTransformation(loopEffect);
      }
    }
    if (zoompan !== false) {
      overrides.format = "auto:animated";
    }
    return {
      options: overrides
    };
  }
});

// src/plugins/cropping.ts
import { normalizeNumberParameter as normalizeNumberParameter2 } from "@cloudinary-util/util";
var cropsAspectRatio = ["auto", "crop", "fill", "lfill", "fill_pad", "thumb"];
var cropsGravityAuto = ["auto", "crop", "fill", "lfill", "fill_pad", "thumb"];
var cropsWithZoom = ["crop", "thumb"];
var DEFAULT_CROP = "limit";
var CroppingPlugin = /* @__PURE__ */ plugin({
  name: "Cropping",
  supports: "all",
  inferOwnOptions: {},
  props: {
    aspectRatio: true,
    crop: true,
    gravity: true,
    zoom: true,
    height: true,
    width: true
  },
  // crop is applied even if the crop key is undefined
  apply: (asset, opts) => {
    let crops = [];
    if (typeof opts.crop === "string" || typeof opts.crop === "undefined") {
      crops.push({
        aspectRatio: opts.aspectRatio,
        height: opts.height,
        gravity: opts.gravity,
        type: opts.crop || DEFAULT_CROP,
        width: opts.width,
        zoom: opts.zoom
      });
    } else if (typeof opts.crop === "object" && !isArray(opts.crop)) {
      crops.push(opts.crop);
    } else if (isArray(opts.crop)) {
      crops = [...opts.crop];
    }
    if (crops.length === 1 && crops[0].source === true) {
      crops.push({
        aspectRatio: opts.aspectRatio,
        width: opts.width,
        height: opts.height,
        gravity: opts.gravity,
        type: DEFAULT_CROP,
        zoom: opts.zoom
      });
    }
    const finalTransformations = [];
    const sourceTransformations = [];
    for (const crop of crops) {
      const cropDimensions = {
        width: crop.width,
        height: crop.height
      };
      if (typeof cropDimensions.width === "undefined" && typeof crop.aspectRatio === "undefined") {
        cropDimensions.width = opts.width;
        if (typeof cropDimensions.height === "undefined") {
          cropDimensions.height = opts.height;
        }
      }
      const transformations = collectTransformations({
        aspectRatio: crop.aspectRatio,
        gravity: crop.gravity,
        type: crop.type || DEFAULT_CROP,
        x: crop.x,
        y: crop.y,
        zoom: crop.zoom,
        ...cropDimensions
      });
      if (transformations.length > 0) {
        if (crop.source === true) {
          sourceTransformations.push(transformations);
        } else {
          finalTransformations.push(transformations);
        }
      }
    }
    sourceTransformations.forEach((transformation) => {
      if (transformation.length > 0) {
        asset.addTransformation(transformation.join(","));
      }
    });
    const results = {
      options: {}
    };
    if (results.options && finalTransformations.length > 0) {
      results.options.resize = finalTransformations.map((transformation) => transformation.join(",")).join("/");
    }
    return results;
  }
});
function collectTransformations(collectOptions) {
  const { aspectRatio, type: crop, x, y, zoom } = collectOptions;
  let gravity = collectOptions.gravity;
  const height = normalizeNumberParameter2(collectOptions.height);
  const width = normalizeNumberParameter2(collectOptions.width);
  const transformations = [];
  const hasDefinedDimensions = height || width;
  const hasValidAspectRatio = aspectRatio && cropsAspectRatio.includes(crop);
  const hasXCoordinate = typeof x === "number" || typeof x === "string";
  const hasYCoordinate = typeof y === "number" || typeof y === "string";
  const hasDefinedCoordinates = hasXCoordinate || hasYCoordinate;
  if (crop && (hasDefinedDimensions || hasValidAspectRatio || hasDefinedCoordinates)) {
    transformations.push(`c_${crop}`);
  }
  if (hasValidAspectRatio) {
    transformations.push(`ar_${aspectRatio}`);
  }
  if (width) {
    transformations.push(`w_${width}`);
  }
  if (!["limit"].includes(crop) && typeof height === "number") {
    transformations.push(`h_${height}`);
  }
  if (hasXCoordinate) {
    transformations.push(`x_${x}`);
  }
  if (hasYCoordinate) {
    transformations.push(`y_${y}`);
  }
  if (!gravity && cropsGravityAuto.includes(crop) && !hasDefinedCoordinates) {
    gravity = "auto";
  }
  if (gravity) {
    if (gravity === "auto" && !cropsGravityAuto.includes(crop)) {
      console.warn(
        `Auto gravity can only be used with crop modes: ${cropsGravityAuto.join(
          ", "
        )}. Not applying gravity.`
      );
    } else {
      transformations.push(`g_${gravity}`);
    }
  }
  if (zoom) {
    if (zoom === "auto" && !cropsWithZoom.includes(crop)) {
      console.warn(
        `Zoom can only be used with crop modes: ${cropsWithZoom.join(
          ", "
        )}. Not applying zoom.`
      );
    } else {
      transformations.push(`z_${zoom}`);
    }
  }
  return transformations;
}

// src/plugins/enhance.ts
var EnhancePlugin = /* @__PURE__ */ plugin({
  name: "Enhance",
  supports: "image",
  inferOwnOptions: {},
  props: {
    enhance: true
  },
  apply: (cldAsset, opts) => {
    if (opts.enhance) {
      cldAsset.addTransformation("e_enhance");
    }
    return {};
  }
});

// src/plugins/extract.ts
var ExtractPlugin = /* @__PURE__ */ plugin({
  name: "Extract",
  supports: "image",
  inferOwnOptions: {},
  props: {
    extract: true
  },
  apply: (cldAsset, { extract }) => {
    const properties = [];
    if (typeof extract === "string") {
      properties.push(`prompt_${extract}`);
    } else if (isArray(extract)) {
      properties.push(`prompt_${formatPrompts(extract)}`);
    } else {
      const prompt = formatPrompts(extract.prompt);
      if (prompt) {
        properties.push(`prompt_${prompt}`);
      }
      if (extract.invert === true) {
        properties.push("invert_true");
      }
      if (typeof extract.mode === "string") {
        properties.push(`mode_${extract.mode}`);
      }
      if (extract.multiple === true) {
        properties.push("multiple_true");
      }
    }
    if (properties.length > 0) {
      const transformation = `e_extract:${properties.join(";")}`;
      cldAsset.addTransformation(transformation);
    }
    return {};
  }
});
function formatPrompts(prompts) {
  if (typeof prompts === "string") return prompts;
  if (isArray(prompts)) {
    return `(${prompts.filter((prompt) => typeof prompt === "string").join(";")})`;
  }
  return void 0;
}

// src/plugins/named-transformations.ts
var NamedTransformationsPlugin = /* @__PURE__ */ plugin({
  name: "NamedTransformations",
  strict: true,
  supports: "all",
  inferOwnOptions: {},
  props: {
    namedTransformations: true,
    transformations: true
  },
  apply: (cldAsset, opts) => {
    const { transformations, namedTransformations } = opts;
    if (transformations && process.env.NODE_ENVIRONMENT === "development") {
      console.warn(
        "The transformations prop is deprecated. Please use namedTransformations instead."
      );
    }
    let _namedTransformations = namedTransformations || transformations || [];
    if (!isArray(_namedTransformations)) {
      _namedTransformations = [_namedTransformations];
    }
    _namedTransformations.forEach((transformation) => {
      cldAsset.addTransformation(`t_${transformation}`);
    });
    return {};
  }
});

// src/plugins/overlays.ts
import { encodeBase64, objectHasKey as objectHasKey2, sortByKey } from "@cloudinary-util/util";
var DEFAULT_TEXT_OPTIONS = {
  color: "black",
  fontFamily: "Arial",
  fontSize: 200,
  fontWeight: "bold"
};
var OverlaysPlugin = /* @__PURE__ */ plugin({
  name: "Overlays",
  supports: "all",
  inferOwnOptions: {},
  props: {
    overlays: true,
    text: true
  },
  apply: (cldAsset, opts) => {
    const { text: text2, overlays = [] } = opts;
    const type = "overlay";
    const typeQualifier = "l";
    if (isArray(overlays)) {
      overlays.forEach(applyOverlay);
    }
    if (typeof text2 === "string") {
      applyOverlay({
        text: Object.assign({}, DEFAULT_TEXT_OPTIONS, {
          text: text2
        })
      });
    } else if (typeof text2 === "object") {
      applyOverlay({
        text: Object.assign({}, DEFAULT_TEXT_OPTIONS, text2)
      });
    }
    function applyOverlay({
      publicId,
      url,
      position: position2,
      text: text3,
      effects: layerEffects = [],
      appliedEffects = [],
      flags: layerFlags = [],
      appliedFlags = [],
      ...options
    }) {
      const hasPublicId = typeof publicId === "string";
      const hasUrl = typeof url === "string";
      const hasText = typeof text3 === "object" || typeof text3 === "string";
      const hasPosition = typeof position2 === "object";
      if (!hasPublicId && !hasUrl && !hasText) {
        console.warn(`An ${type} is missing Public ID, URL, or Text`);
        return;
      }
      let layerTransformation;
      if (hasText) {
        layerTransformation = `${typeQualifier}_text`;
      } else if (hasPublicId) {
        layerTransformation = `${typeQualifier}_${publicId.replace(
          /\//g,
          ":"
        )}`;
      } else if (hasUrl) {
        layerTransformation = `${typeQualifier}_fetch:${encodeBase64(url)}`;
      }
      const primary2 = [];
      const applied = [];
      Object.keys(options).forEach((key) => {
        if (!objectHasKey2(primary, key)) return;
        const { qualifier, converters } = primary[key];
        const transformation = constructTransformation({
          qualifier,
          value: options[key],
          converters
        });
        if (transformation) {
          primary2.push(transformation);
        }
      });
      layerEffects.forEach((effect) => {
        Object.keys(effect).forEach((key) => {
          const effectQualifier = primary[key] || effects[key] || position[key];
          if (!effectQualifier) return;
          const { qualifier, prefix, converters } = effectQualifier;
          const transformation = constructTransformation({
            qualifier,
            prefix,
            value: effect[key],
            converters
          });
          if (transformation) {
            primary2.push(transformation);
          }
        });
      });
      appliedEffects.forEach((effect) => {
        Object.keys(effect).forEach((key) => {
          const effectQualifier = primary[key] || effects[key] || position[key];
          if (!effectQualifier) return;
          const { qualifier, prefix, converters } = effectQualifier;
          const transformation = constructTransformation({
            qualifier,
            prefix,
            value: effect[key],
            converters
          });
          if (transformation) {
            applied.push(transformation);
          }
        });
      });
      const activeLayerFlags = isArray(layerFlags) ? layerFlags : [layerFlags];
      activeLayerFlags.forEach((flag) => {
        primary2.push(`fl_${flag}`);
      });
      const activeAppliedFlags = isArray(appliedFlags) ? appliedFlags : [appliedFlags];
      activeAppliedFlags.forEach((flag) => {
        applied.push(`fl_${flag}`);
      });
      if (hasText) {
        if (typeof text3 === "string") {
          text3 = {
            ...DEFAULT_TEXT_OPTIONS,
            text: text3
          };
        }
        const textTransformations = [];
        if (typeof text3 === "object") {
          const textOptions = entriesOf(text3).flatMap(
            ([key, value]) => {
              if (!text[key]) return [];
              return {
                ...text[key],
                key,
                value,
                order: text[key].order || 99
              };
            }
          );
          const sortedTextOptions = sortByKey(textOptions, "order");
          for (const textOption of sortedTextOptions) {
            const { key, value, qualifier, location, converters } = textOption;
            let textValue = value;
            converters?.forEach(({ test, convert }) => {
              if (!test(value)) return;
              textValue = convert(value);
            });
            if (location === "primary") {
              primary2.push(`${qualifier}_${textValue}`);
            } else if (qualifier === "self") {
              textTransformations.push(key);
            } else if (qualifier) {
              textTransformations.push(`${qualifier}_${textValue}`);
            } else {
              textTransformations.push(textValue);
            }
          }
        }
        const specialCharacters = {
          ".": "%2E",
          ",": "%2C",
          "/": "%2F"
        };
        let layerText = text3?.text || "";
        if (typeof layerText === "string") {
          Object.keys(specialCharacters)?.forEach((character) => {
            layerText = layerText?.replace(
              character,
              specialCharacters[character]
            );
          });
        }
        layerTransformation = `${layerTransformation}:${textTransformations.join(
          "_"
        )}:${layerText}`;
      }
      if (hasPosition) {
        Object.entries(position2).forEach(([key, value]) => {
          if (!objectHasKey2(position, key)) return;
          const { qualifier, converters } = position[key];
          const transformation = constructTransformation({
            qualifier,
            value,
            converters
          });
          if (transformation) {
            applied.push(transformation);
          }
        });
      }
      if (primary2.length > 0) {
        layerTransformation = `${layerTransformation},${primary2.join(",")}`;
      }
      layerTransformation = `${layerTransformation}/fl_layer_apply,fl_no_overflow`;
      if (applied.length > 0) {
        layerTransformation = `${layerTransformation},${applied.join(",")}`;
      }
      cldAsset.addTransformation(layerTransformation);
    }
    return {};
  }
});

// src/plugins/version.ts
var VersionPlugin = /* @__PURE__ */ plugin({
  name: "Version",
  supports: "all",
  inferOwnOptions: {},
  props: {
    version: true
  },
  apply: (cldAsset, opts) => {
    const { version } = opts;
    if (typeof version !== "string" && typeof version !== "number") return {};
    cldAsset.setVersion(`${version}`.replace("v", ""));
    return {};
  }
});

// src/lib/cloudinary.ts
var transformationPlugins = [
  // Some features *must* be the first transformation applied
  // thus their plugins *must* come first in the chain
  EnhancePlugin,
  ExtractPlugin,
  RecolorPlugin,
  RemoveBackgroundPlugin,
  RemovePlugin,
  ReplacePlugin,
  ReplaceBackgroundPlugin,
  RestorePlugin,
  // Cropping needs to be before any other general transformations
  // as it provides the option of 2-step resizing where someone
  // can resize the "base" canvas as well as the final resize
  // mechanism commonly used for responsive resizing
  CroppingPlugin,
  // Raw transformations should always come before
  // other arguments to avoid conflicting with
  // added options via the component
  PreserveTransformationsPlugin,
  RawTransformationsPlugin,
  AbrPlugin,
  DefaultImagePlugin,
  EffectsPlugin,
  FillBackgroundPlugin,
  FlagsPlugin,
  OverlaysPlugin,
  SanitizePlugin,
  NamedTransformationsPlugin,
  SeoPlugin,
  UnderlaysPlugin,
  VersionPlugin,
  ZoompanPlugin
];
var getCloudinaryPluginProps = (plugins) => {
  const cloudinaryPluginProps2 = {};
  plugins.forEach((plugin2) => {
    Object.assign(cloudinaryPluginProps2, plugin2.props);
  });
  return cloudinaryPluginProps2;
};
var cloudinaryPluginProps = /* @__PURE__ */ getCloudinaryPluginProps(transformationPlugins);
var cloudinaryPluginKeys = /* @__PURE__ */ Object.keys(
  cloudinaryPluginProps
);
function constructCloudinaryUrl({ options, config = {}, analytics }) {
  if (analytics === false) {
    if (typeof config?.url === "undefined") {
      config.url = {};
    }
    config.url.analytics = false;
  }
  const cld = new Cloudinary(config);
  if (typeof options?.src !== "string") {
    throw Error(
      `Failed to construct Cloudinary URL: Missing source (src) in options.`
    );
  }
  if (!options?.assetType) {
    options.assetType = "image";
  }
  const parsedOptions = {};
  let publicId;
  if (typeof options.src === "string" && /^https?:\/\//.test(options.src)) {
    try {
      const parts = parseUrl(options.src);
      publicId = parts?.publicId;
      parsedOptions.seoSuffix = parts?.seoSuffix;
      parsedOptions.version = parts?.version;
    } catch (e) {
    }
  }
  if (!publicId) {
    publicId = options.src;
  }
  entriesOf(parsedOptions).forEach(([key, value]) => {
    if (objectHasKey3(options, key)) return;
    options[key] = value;
  });
  options.version ?? (options.version = 1);
  const normalizedAssetType = options.assetType === "image" || options.assetType === "images" ? "image" : options.assetType === "video" || options.assetType === "videos" ? "video" : throwError(`${options.assetType} is not a valid assetType`);
  const cldAsset = cld[normalizedAssetType](publicId);
  const pluginEffects = {};
  transformationPlugins.forEach(
    ({ name, apply, strict, supports, props }) => {
      const shouldApply = Object.keys(props).some(
        (key) => options[key] !== void 0
      );
      if (!shouldApply) return;
      if (normalizedAssetType !== supports && supports !== "all") {
        console.warn(
          `${name} does not support assetType ${normalizedAssetType}`
        );
        return;
      }
      if (options.strictTransformations && !strict) {
        console.warn(`${name} does not support Strict Transformations.`);
        return;
      }
      const results = apply(cldAsset, options);
      const pluginOptions = results?.options ?? {};
      Object.assign(pluginEffects, pluginOptions);
    }
  );
  if (typeof pluginEffects.resize === "string") {
    cldAsset.addTransformation(pluginEffects.resize);
  }
  cldAsset.setDeliveryType(options?.deliveryType || "upload");
  if (!options.strictTransformations) {
    if (options?.dpr) {
      let dpr = options.dpr;
      if (typeof dpr === "number") {
        dpr = dpr.toFixed(1);
      }
      cldAsset.addTransformation(`dpr_${dpr}`);
    }
    const defaultFormat = options?.format === "default";
    const rawContainsFormat = searchAssetRawTransformations("f_", cldAsset, {
      matchType: "startsWith"
    });
    const rawContainsFormatAndExplicit = rawContainsFormat && typeof options?.format !== "undefined";
    if (pluginEffects?.format || !defaultFormat && (!rawContainsFormat || rawContainsFormatAndExplicit)) {
      cldAsset.format(options?.format || pluginEffects?.format || "auto");
    }
    const defaultQuality = options?.quality === "default";
    const rawContainsQuality = searchAssetRawTransformations("q_", cldAsset, {
      matchType: "startsWith"
    });
    const rawContainsQualityAndExplicit = rawContainsQuality && typeof options?.quality !== "undefined";
    if (!defaultQuality && (!rawContainsQuality || rawContainsQualityAndExplicit)) {
      cldAsset.quality(options?.quality || "auto");
    }
  }
  return cldAsset.toURL({
    trackedAnalytics: typeof analytics === "object" ? analytics : void 0
  });
}
function searchAssetRawTransformations(query, asset, options) {
  if (typeof asset.transformation === "undefined") return;
  const { matchType = "includes" } = options || {};
  const transformations = asset.transformation.actions.flatMap(
    (transformation) => {
      return transformation.toString().split("/").flatMap((seg) => seg.split(","));
    }
  );
  const matches = transformations.filter((transformation) => {
    if (matchType === "startsWith") {
      return transformation.startsWith(query);
    } else {
      return transformation.includes(query);
    }
  });
  return matches.length > 0;
}
export {
  cloudinaryPluginKeys,
  cloudinaryPluginProps,
  constructCloudinaryUrl,
  searchAssetRawTransformations,
  transformationPlugins
};
