// @cloudinary-util/url-loader/dist/lib/upload-widget.js
// src/lib/upload-widget.ts
function getUploadWidgetOptions({ uploadSignature, ...options }, config) {
  const signed = typeof uploadSignature === "function";
  const { cloudName, apiKey } = config?.cloud || {};
  if (!cloudName) {
    throw new Error(
      "A Cloudinary Cloud name is required, please make sure your environment variable is set and configured in your environment."
    );
  }
  if (signed && !apiKey) {
    throw new Error(
      "A Cloudinary API Key is required for signed requests, please make sure your environment variable is set and configured in your environment."
    );
  }
  if (!signed && !options.uploadPreset) {
    throw new Error(
      "A Cloudinary Upload Preset is required for unsigned uploads. Please specify an uploadPreset or configure signed uploads."
    );
  }
  const uploadOptions = {
    cloudName,
    apiKey,
    ...options
  };
  if (signed) {
    uploadOptions.uploadSignature = uploadSignature;
  }
  return uploadOptions;
}
var UPLOAD_WIDGET_EVENTS = {
  abort: "onAbort",
  "batch-cancelled": "onBatchCancelled",
  close: "onClose",
  "display-changed": "onDisplayChanged",
  publicid: "onPublicId",
  "queues-end": "onQueuesEnd",
  "queues-start": "onQueuesStart",
  retry: "onRetry",
  "show-completed": "onShowCompleted",
  "source-changed": "onSourceChanged",
  success: "onSuccess",
  tags: "onTags",
  "upload-added": "onUploadAdded"
};
function generateUploadWidgetResultCallback(options) {
  return function resultCallback(error, uploadResult) {
    if (error) {
      if (typeof options.onError === "function") {
        options.onError(error, uploadResult);
      }
    }
    if (typeof options.onResult === "function") {
      options.onResult(uploadResult);
    }
    const widgetEvent = typeof uploadResult?.event === "string" && UPLOAD_WIDGET_EVENTS[uploadResult.event];
    if (typeof widgetEvent === "string" && typeof options[widgetEvent] === "function") {
      const callback = options[widgetEvent];
      callback(uploadResult);
    }
  };
}
export {
  UPLOAD_WIDGET_EVENTS,
  generateUploadWidgetResultCallback,
  getUploadWidgetOptions
};
