// @cloudinary-util/url-loader/dist/lib/upload.cjs
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/lib/upload.ts
var upload_exports = {};
__export(upload_exports, {
  generateSignatureCallback: () => generateSignatureCallback
});
module.exports = __toCommonJS(upload_exports);
function generateSignatureCallback({
  signatureEndpoint,
  fetch: fetcher
}) {
  return function generateSignature(callback, paramsToSign) {
    if (typeof signatureEndpoint === "undefined") {
      throw Error(
        "Failed to generate signature: signatureEndpoint property undefined."
      );
    }
    if (typeof fetcher === "undefined") {
      throw Error("Failed to generate signature: fetch property undefined.");
    }
    fetcher(signatureEndpoint, {
      method: "POST",
      body: JSON.stringify({
        paramsToSign
      }),
      headers: {
        "Content-Type": "application/json"
      }
    }).then((response) => response.json()).then((result) => {
      callback(result.signature);
    }).catch((error) => {
      callback(null, error);
    });
  };
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  generateSignatureCallback
});
