// @cloudinary-util/url-loader/dist/lib/upload-widget.cjs
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/lib/upload-widget.ts
var upload_widget_exports = {};
__export(upload_widget_exports, {
  UPLOAD_WIDGET_EVENTS: () => UPLOAD_WIDGET_EVENTS,
  generateUploadWidgetResultCallback: () => generateUploadWidgetResultCallback,
  getUploadWidgetOptions: () => getUploadWidgetOptions
});
module.exports = __toCommonJS(upload_widget_exports);
function getUploadWidgetOptions({ uploadSignature, ...options }, config) {
  const signed = typeof uploadSignature === "function";
  const { cloudName, apiKey } = config?.cloud || {};
  if (!cloudName) {
    throw new Error(
      "A Cloudinary Cloud name is required, please make sure your environment variable is set and configured in your environment."
    );
  }
  if (signed && !apiKey) {
    throw new Error(
      "A Cloudinary API Key is required for signed requests, please make sure your environment variable is set and configured in your environment."
    );
  }
  if (!signed && !options.uploadPreset) {
    throw new Error(
      "A Cloudinary Upload Preset is required for unsigned uploads. Please specify an uploadPreset or configure signed uploads."
    );
  }
  const uploadOptions = {
    cloudName,
    apiKey,
    ...options
  };
  if (signed) {
    uploadOptions.uploadSignature = uploadSignature;
  }
  return uploadOptions;
}
var UPLOAD_WIDGET_EVENTS = {
  abort: "onAbort",
  "batch-cancelled": "onBatchCancelled",
  close: "onClose",
  "display-changed": "onDisplayChanged",
  publicid: "onPublicId",
  "queues-end": "onQueuesEnd",
  "queues-start": "onQueuesStart",
  retry: "onRetry",
  "show-completed": "onShowCompleted",
  "source-changed": "onSourceChanged",
  success: "onSuccess",
  tags: "onTags",
  "upload-added": "onUploadAdded"
};
function generateUploadWidgetResultCallback(options) {
  return function resultCallback(error, uploadResult) {
    if (error) {
      if (typeof options.onError === "function") {
        options.onError(error, uploadResult);
      }
    }
    if (typeof options.onResult === "function") {
      options.onResult(uploadResult);
    }
    const widgetEvent = typeof uploadResult?.event === "string" && UPLOAD_WIDGET_EVENTS[uploadResult.event];
    if (typeof widgetEvent === "string" && typeof options[widgetEvent] === "function") {
      const callback = options[widgetEvent];
      callback(uploadResult);
    }
  };
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  UPLOAD_WIDGET_EVENTS,
  generateUploadWidgetResultCallback,
  getUploadWidgetOptions
});
