// @cloudinary-util/url-loader/dist/plugins/abr.cjs
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/plugins/abr.ts
var abr_exports = {};
__export(abr_exports, {
  AbrPlugin: () => AbrPlugin
});
module.exports = __toCommonJS(abr_exports);

// src/lib/plugin.ts
var plugin = (def) => ({
  strict: false,
  alwaysApply: false,
  ...def,
  apply: (cldAsset, ctx) => def.apply(cldAsset, ctx, ctx)
});

// src/plugins/abr.ts
var AbrPlugin = /* @__PURE__ */ plugin({
  name: "Abr",
  supports: "video",
  inferOwnOptions: {},
  props: {
    streamingProfile: true
  },
  apply: (asset, opts) => {
    if (typeof opts.streamingProfile !== "string") return {};
    asset.addTransformation(`sp_${opts.streamingProfile}`);
    return {};
  }
});
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  AbrPlugin
});
