// @cloudinary-util/url-loader/dist/plugins/cropping.cjs
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/plugins/cropping.ts
var cropping_exports = {};
__export(cropping_exports, {
  CroppingPlugin: () => CroppingPlugin
});
module.exports = __toCommonJS(cropping_exports);
var import_util = require("@cloudinary-util/util");

// src/lib/plugin.ts
var plugin = (def) => ({
  strict: false,
  alwaysApply: false,
  ...def,
  apply: (cldAsset, ctx) => def.apply(cldAsset, ctx, ctx)
});

// src/lib/utils.ts
var isArray = Array.isArray;

// src/plugins/cropping.ts
var cropsAspectRatio = ["auto", "crop", "fill", "lfill", "fill_pad", "thumb"];
var cropsGravityAuto = ["auto", "crop", "fill", "lfill", "fill_pad", "thumb"];
var cropsWithZoom = ["crop", "thumb"];
var DEFAULT_CROP = "limit";
var CroppingPlugin = /* @__PURE__ */ plugin({
  name: "Cropping",
  supports: "all",
  inferOwnOptions: {},
  props: {
    aspectRatio: true,
    crop: true,
    gravity: true,
    zoom: true,
    height: true,
    width: true
  },
  // crop is applied even if the crop key is undefined
  apply: (asset, opts) => {
    let crops = [];
    if (typeof opts.crop === "string" || typeof opts.crop === "undefined") {
      crops.push({
        aspectRatio: opts.aspectRatio,
        height: opts.height,
        gravity: opts.gravity,
        type: opts.crop || DEFAULT_CROP,
        width: opts.width,
        zoom: opts.zoom
      });
    } else if (typeof opts.crop === "object" && !isArray(opts.crop)) {
      crops.push(opts.crop);
    } else if (isArray(opts.crop)) {
      crops = [...opts.crop];
    }
    if (crops.length === 1 && crops[0].source === true) {
      crops.push({
        aspectRatio: opts.aspectRatio,
        width: opts.width,
        height: opts.height,
        gravity: opts.gravity,
        type: DEFAULT_CROP,
        zoom: opts.zoom
      });
    }
    const finalTransformations = [];
    const sourceTransformations = [];
    for (const crop of crops) {
      const cropDimensions = {
        width: crop.width,
        height: crop.height
      };
      if (typeof cropDimensions.width === "undefined" && typeof crop.aspectRatio === "undefined") {
        cropDimensions.width = opts.width;
        if (typeof cropDimensions.height === "undefined") {
          cropDimensions.height = opts.height;
        }
      }
      const transformations = collectTransformations({
        aspectRatio: crop.aspectRatio,
        gravity: crop.gravity,
        type: crop.type || DEFAULT_CROP,
        x: crop.x,
        y: crop.y,
        zoom: crop.zoom,
        ...cropDimensions
      });
      if (transformations.length > 0) {
        if (crop.source === true) {
          sourceTransformations.push(transformations);
        } else {
          finalTransformations.push(transformations);
        }
      }
    }
    sourceTransformations.forEach((transformation) => {
      if (transformation.length > 0) {
        asset.addTransformation(transformation.join(","));
      }
    });
    const results = {
      options: {}
    };
    if (results.options && finalTransformations.length > 0) {
      results.options.resize = finalTransformations.map((transformation) => transformation.join(",")).join("/");
    }
    return results;
  }
});
function collectTransformations(collectOptions) {
  const { aspectRatio, type: crop, x, y, zoom } = collectOptions;
  let gravity = collectOptions.gravity;
  const height = (0, import_util.normalizeNumberParameter)(collectOptions.height);
  const width = (0, import_util.normalizeNumberParameter)(collectOptions.width);
  const transformations = [];
  const hasDefinedDimensions = height || width;
  const hasValidAspectRatio = aspectRatio && cropsAspectRatio.includes(crop);
  const hasXCoordinate = typeof x === "number" || typeof x === "string";
  const hasYCoordinate = typeof y === "number" || typeof y === "string";
  const hasDefinedCoordinates = hasXCoordinate || hasYCoordinate;
  if (crop && (hasDefinedDimensions || hasValidAspectRatio || hasDefinedCoordinates)) {
    transformations.push(`c_${crop}`);
  }
  if (hasValidAspectRatio) {
    transformations.push(`ar_${aspectRatio}`);
  }
  if (width) {
    transformations.push(`w_${width}`);
  }
  if (!["limit"].includes(crop) && typeof height === "number") {
    transformations.push(`h_${height}`);
  }
  if (hasXCoordinate) {
    transformations.push(`x_${x}`);
  }
  if (hasYCoordinate) {
    transformations.push(`y_${y}`);
  }
  if (!gravity && cropsGravityAuto.includes(crop) && !hasDefinedCoordinates) {
    gravity = "auto";
  }
  if (gravity) {
    if (gravity === "auto" && !cropsGravityAuto.includes(crop)) {
      console.warn(
        `Auto gravity can only be used with crop modes: ${cropsGravityAuto.join(
          ", "
        )}. Not applying gravity.`
      );
    } else {
      transformations.push(`g_${gravity}`);
    }
  }
  if (zoom) {
    if (zoom === "auto" && !cropsWithZoom.includes(crop)) {
      console.warn(
        `Zoom can only be used with crop modes: ${cropsWithZoom.join(
          ", "
        )}. Not applying zoom.`
      );
    } else {
      transformations.push(`z_${zoom}`);
    }
  }
  return transformations;
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  CroppingPlugin
});
