// @cloudinary-util/url-loader/dist/plugins/default-image.cjs
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/plugins/default-image.ts
var default_image_exports = {};
__export(default_image_exports, {
  DefaultImagePlugin: () => DefaultImagePlugin
});
module.exports = __toCommonJS(default_image_exports);
var import_util = require("@cloudinary-util/util");

// src/lib/plugin.ts
var plugin = (def) => ({
  strict: false,
  alwaysApply: false,
  ...def,
  apply: (cldAsset, ctx) => def.apply(cldAsset, ctx, ctx)
});

// src/plugins/default-image.ts
var DefaultImagePlugin = /* @__PURE__ */ plugin({
  name: "DefaultImage",
  supports: "image",
  inferOwnOptions: {},
  props: { defaultImage: true },
  apply: (asset, opts) => {
    const { defaultImage } = opts;
    if (typeof defaultImage !== "string") return {};
    if (!(0, import_util.getFormat)(defaultImage)) {
      console.warn(
        `The defaultImage prop may be missing a format and must include it along with the public ID. (Ex: myimage.jpg)`
      );
    }
    const defaultImageId = defaultImage.replace(/\//g, ":");
    asset.addTransformation(`d_${defaultImageId}`);
    return {};
  }
});
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  DefaultImagePlugin
});
