// @cloudinary-util/url-loader/dist/plugins/extract.js
// src/lib/plugin.ts
var plugin = (def) => ({
  strict: false,
  alwaysApply: false,
  ...def,
  apply: (cldAsset, ctx) => def.apply(cldAsset, ctx, ctx)
});

// src/lib/utils.ts
var isArray = Array.isArray;

// src/plugins/extract.ts
var ExtractPlugin = /* @__PURE__ */ plugin({
  name: "Extract",
  supports: "image",
  inferOwnOptions: {},
  props: {
    extract: true
  },
  apply: (cldAsset, { extract }) => {
    const properties = [];
    if (typeof extract === "string") {
      properties.push(`prompt_${extract}`);
    } else if (isArray(extract)) {
      properties.push(`prompt_${formatPrompts(extract)}`);
    } else {
      const prompt = formatPrompts(extract.prompt);
      if (prompt) {
        properties.push(`prompt_${prompt}`);
      }
      if (extract.invert === true) {
        properties.push("invert_true");
      }
      if (typeof extract.mode === "string") {
        properties.push(`mode_${extract.mode}`);
      }
      if (extract.multiple === true) {
        properties.push("multiple_true");
      }
    }
    if (properties.length > 0) {
      const transformation = `e_extract:${properties.join(";")}`;
      cldAsset.addTransformation(transformation);
    }
    return {};
  }
});
function formatPrompts(prompts) {
  if (typeof prompts === "string") return prompts;
  if (isArray(prompts)) {
    return `(${prompts.filter((prompt) => typeof prompt === "string").join(";")})`;
  }
  return void 0;
}
export {
  ExtractPlugin
};
