// @cloudinary-util/url-loader/dist/plugins/fill-background.cjs
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/plugins/fill-background.ts
var fill_background_exports = {};
__export(fill_background_exports, {
  FillBackgroundPlugin: () => FillBackgroundPlugin
});
module.exports = __toCommonJS(fill_background_exports);
var import_util = require("@cloudinary-util/util");

// src/lib/plugin.ts
var plugin = (def) => ({
  strict: false,
  alwaysApply: false,
  ...def,
  apply: (cldAsset, ctx) => def.apply(cldAsset, ctx, ctx)
});

// src/plugins/fill-background.ts
var defaultCrop = "pad";
var FillBackgroundPlugin = /* @__PURE__ */ plugin({
  name: "FillBackground",
  supports: "image",
  inferOwnOptions: {},
  props: {
    fillBackground: true
  },
  apply: (cldAsset, opts, ctx) => {
    const { fillBackground } = opts;
    if (typeof fillBackground === "undefined") return {};
    const width = (0, import_util.normalizeNumberParameter)(ctx.width);
    const height = (0, import_util.normalizeNumberParameter)(ctx.height);
    const hasDefinedDimensions = typeof height === "number" && typeof width === "number";
    let aspectRatio = ctx.aspectRatio;
    if (!aspectRatio && hasDefinedDimensions) {
      aspectRatio = `${width}:${height}`;
    }
    if (!aspectRatio) {
      if (process.env.NODE_ENV === "development") {
        console.warn(
          `Could not determine aspect ratio based on available options to use fillBackground. Please specify width and height or an aspect ratio.`
        );
      }
      return {};
    }
    if (fillBackground === true) {
      const properties = [
        "b_gen_fill",
        `ar_${aspectRatio}`,
        `c_${defaultCrop}`
      ];
      cldAsset.addTransformation(properties.join(","));
    } else if (typeof fillBackground === "object") {
      const { crop = defaultCrop, gravity, prompt } = fillBackground;
      const properties = [`ar_${aspectRatio}`, `c_${crop}`];
      if (typeof prompt === "string") {
        properties.unshift(`b_gen_fill:${prompt}`);
      } else {
        properties.unshift(`b_gen_fill`);
      }
      if (typeof gravity === "string") {
        properties.push(`g_${gravity}`);
      }
      cldAsset.addTransformation(properties.join(","));
    }
    return {};
  }
});
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  FillBackgroundPlugin
});
