// @cloudinary-util/url-loader/dist/plugins/flags.cjs
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/plugins/flags.ts
var flags_exports = {};
__export(flags_exports, {
  FlagsPlugin: () => FlagsPlugin
});
module.exports = __toCommonJS(flags_exports);

// src/lib/plugin.ts
var plugin = (def) => ({
  strict: false,
  alwaysApply: false,
  ...def,
  apply: (cldAsset, ctx) => def.apply(cldAsset, ctx, ctx)
});

// src/lib/utils.ts
var isArray = Array.isArray;

// src/plugins/flags.ts
var FlagsPlugin = /* @__PURE__ */ plugin({
  name: "Flags",
  supports: "all",
  inferOwnOptions: {},
  props: {
    flags: true
  },
  apply: (cldAsset, { flags }) => {
    if (typeof flags === "string") {
      flags = [flags];
    }
    if (isArray(flags)) {
      flags.forEach((flag) => cldAsset.addFlag(flag));
    } else if (typeof flags === "object") {
      Object.entries(flags).forEach(([qualifier, value]) => {
        cldAsset.addTransformation(`fl_${qualifier}:${value}`);
      });
    }
    return {};
  }
});
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  FlagsPlugin
});
