// @cloudinary-util/url-loader/dist/plugins/named-transformations.js
// src/lib/plugin.ts
var plugin = (def) => ({
  strict: false,
  alwaysApply: false,
  ...def,
  apply: (cldAsset, ctx) => def.apply(cldAsset, ctx, ctx)
});

// src/lib/utils.ts
var isArray = Array.isArray;

// src/plugins/named-transformations.ts
var NamedTransformationsPlugin = /* @__PURE__ */ plugin({
  name: "NamedTransformations",
  strict: true,
  supports: "all",
  inferOwnOptions: {},
  props: {
    namedTransformations: true,
    transformations: true
  },
  apply: (cldAsset, opts) => {
    const { transformations, namedTransformations } = opts;
    if (transformations && process.env.NODE_ENVIRONMENT === "development") {
      console.warn(
        "The transformations prop is deprecated. Please use namedTransformations instead."
      );
    }
    let _namedTransformations = namedTransformations || transformations || [];
    if (!isArray(_namedTransformations)) {
      _namedTransformations = [_namedTransformations];
    }
    _namedTransformations.forEach((transformation) => {
      cldAsset.addTransformation(`t_${transformation}`);
    });
    return {};
  }
});
export {
  NamedTransformationsPlugin
};
