// @cloudinary-util/url-loader/dist/plugins/named-transformations.cjs
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/plugins/named-transformations.ts
var named_transformations_exports = {};
__export(named_transformations_exports, {
  NamedTransformationsPlugin: () => NamedTransformationsPlugin
});
module.exports = __toCommonJS(named_transformations_exports);

// src/lib/plugin.ts
var plugin = (def) => ({
  strict: false,
  alwaysApply: false,
  ...def,
  apply: (cldAsset, ctx) => def.apply(cldAsset, ctx, ctx)
});

// src/lib/utils.ts
var isArray = Array.isArray;

// src/plugins/named-transformations.ts
var NamedTransformationsPlugin = /* @__PURE__ */ plugin({
  name: "NamedTransformations",
  strict: true,
  supports: "all",
  inferOwnOptions: {},
  props: {
    namedTransformations: true,
    transformations: true
  },
  apply: (cldAsset, opts) => {
    const { transformations, namedTransformations } = opts;
    if (transformations && process.env.NODE_ENVIRONMENT === "development") {
      console.warn(
        "The transformations prop is deprecated. Please use namedTransformations instead."
      );
    }
    let _namedTransformations = namedTransformations || transformations || [];
    if (!isArray(_namedTransformations)) {
      _namedTransformations = [_namedTransformations];
    }
    _namedTransformations.forEach((transformation) => {
      cldAsset.addTransformation(`t_${transformation}`);
    });
    return {};
  }
});
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  NamedTransformationsPlugin
});
