// @cloudinary-util/url-loader/dist/plugins/preserve-transformations.js
// src/plugins/preserve-transformations.ts
import { getTransformations } from "@cloudinary-util/util";

// src/lib/plugin.ts
var plugin = (def) => ({
  strict: false,
  alwaysApply: false,
  ...def,
  apply: (cldAsset, ctx) => def.apply(cldAsset, ctx, ctx)
});

// src/plugins/preserve-transformations.ts
var PreserveTransformationsPlugin = /* @__PURE__ */ plugin({
  name: "PreserveTransformations",
  supports: "all",
  inferOwnOptions: {},
  props: {
    preserveTransformations: true
  },
  apply: (cldAsset, opts, ctx) => {
    const { preserveTransformations = false } = opts;
    if (preserveTransformations) {
      try {
        if (ctx.src === void 0) {
          throw new Error("options.src was undefined");
        }
        const transformations = getTransformations(ctx.src).map(
          (t) => t.join(",")
        );
        transformations.flat().forEach((transformation) => {
          cldAsset.addTransformation(transformation);
        });
      } catch (e) {
        console.warn(
          `Failed to preserve transformations: ${e.message}`
        );
      }
    }
    return {};
  }
});
export {
  PreserveTransformationsPlugin
};
