// @cloudinary-util/url-loader/dist/plugins/preserve-transformations.cjs
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/plugins/preserve-transformations.ts
var preserve_transformations_exports = {};
__export(preserve_transformations_exports, {
  PreserveTransformationsPlugin: () => PreserveTransformationsPlugin
});
module.exports = __toCommonJS(preserve_transformations_exports);
var import_util = require("@cloudinary-util/util");

// src/lib/plugin.ts
var plugin = (def) => ({
  strict: false,
  alwaysApply: false,
  ...def,
  apply: (cldAsset, ctx) => def.apply(cldAsset, ctx, ctx)
});

// src/plugins/preserve-transformations.ts
var PreserveTransformationsPlugin = /* @__PURE__ */ plugin({
  name: "PreserveTransformations",
  supports: "all",
  inferOwnOptions: {},
  props: {
    preserveTransformations: true
  },
  apply: (cldAsset, opts, ctx) => {
    const { preserveTransformations = false } = opts;
    if (preserveTransformations) {
      try {
        if (ctx.src === void 0) {
          throw new Error("options.src was undefined");
        }
        const transformations = (0, import_util.getTransformations)(ctx.src).map(
          (t) => t.join(",")
        );
        transformations.flat().forEach((transformation) => {
          cldAsset.addTransformation(transformation);
        });
      } catch (e) {
        console.warn(
          `Failed to preserve transformations: ${e.message}`
        );
      }
    }
    return {};
  }
});
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  PreserveTransformationsPlugin
});
