// @cloudinary-util/url-loader/dist/plugins/remove-background.cjs
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/plugins/remove-background.ts
var remove_background_exports = {};
__export(remove_background_exports, {
  RemoveBackgroundPlugin: () => RemoveBackgroundPlugin
});
module.exports = __toCommonJS(remove_background_exports);

// src/lib/plugin.ts
var plugin = (def) => ({
  strict: false,
  alwaysApply: false,
  ...def,
  apply: (cldAsset, ctx) => def.apply(cldAsset, ctx, ctx)
});

// src/plugins/remove-background.ts
var RemoveBackgroundPlugin = /* @__PURE__ */ plugin({
  name: "RemoveBackground",
  supports: "image",
  inferOwnOptions: {},
  props: {
    removeBackground: true
  },
  apply: (cldAsset, opts) => {
    if (opts.removeBackground) {
      cldAsset.addTransformation("e_background_removal");
    }
    return {};
  }
});
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  RemoveBackgroundPlugin
});
