// @cloudinary-util/url-loader/dist/plugins/replace-background.js
// src/lib/plugin.ts
var plugin = (def) => ({
  strict: false,
  alwaysApply: false,
  ...def,
  apply: (cldAsset, ctx) => def.apply(cldAsset, ctx, ctx)
});

// src/plugins/replace-background.ts
var ReplaceBackgroundPlugin = /* @__PURE__ */ plugin({
  name: "ReplaceBackground",
  supports: "image",
  inferOwnOptions: {},
  props: {
    replaceBackground: true
  },
  apply: (cldAsset, opts) => {
    const { replaceBackground } = opts;
    if (!replaceBackground) return {};
    const properties = [];
    if (typeof replaceBackground === "object") {
      if (typeof replaceBackground.prompt !== "undefined") {
        properties.push(`prompt_${replaceBackground.prompt}`);
      }
      if (typeof replaceBackground.seed === "number") {
        properties.push(`seed_${replaceBackground.seed}`);
      }
    } else if (typeof replaceBackground === "string") {
      properties.push(`prompt_${replaceBackground}`);
    }
    let transformation = "e_gen_background_replace";
    if (properties.length > 0) {
      transformation = `${transformation}:${properties.join(";")}`;
    }
    cldAsset.addTransformation(transformation);
    return {};
  }
});
export {
  ReplaceBackgroundPlugin
};
