// @cloudinary-util/url-loader/dist/plugins/replace.js
// src/lib/plugin.ts
var plugin = (def) => ({
  strict: false,
  alwaysApply: false,
  ...def,
  apply: (cldAsset, ctx) => def.apply(cldAsset, ctx, ctx)
});

// src/lib/utils.ts
var isArray = Array.isArray;

// src/plugins/replace.ts
var ReplacePlugin = /* @__PURE__ */ plugin({
  name: "Replace",
  supports: "image",
  inferOwnOptions: {},
  props: {
    replace: true
  },
  apply: (cldAsset, opts) => {
    const { replace } = opts;
    if (!replace) return {};
    let from, to, preserveGeometry = false;
    if (isArray(replace)) {
      from = replace[0];
      to = replace[1];
      preserveGeometry = replace[2] || false;
    } else {
      from = replace.from;
      to = replace.to;
      preserveGeometry = replace.preserveGeometry || false;
    }
    const properties = [`e_gen_replace:from_${from}`, `to_${to}`];
    if (preserveGeometry) {
      properties.push(`preserve-geometry_${preserveGeometry}`);
    }
    cldAsset.addTransformation(properties.join(";"));
    return {};
  }
});
export {
  ReplacePlugin
};
