// @cloudinary-util/url-loader/dist/plugins/underlays.js
// src/plugins/underlays.ts
import { objectHasKey } from "@cloudinary-util/util";

// src/constants/qualifiers.ts
import { convertColorHexToRgb, testColorIsHex } from "@cloudinary-util/util";
var primary = {
  aspectRatio: {
    qualifier: "ar"
  },
  crop: {
    qualifier: "c"
  },
  gravity: {
    qualifier: "g"
  },
  height: {
    qualifier: "h"
  },
  width: { qualifier: "w" }
};
var position = {
  angle: {
    qualifier: "a"
  },
  gravity: {
    qualifier: "g"
  },
  x: { qualifier: "x" },
  y: { qualifier: "y" }
};

// src/lib/plugin.ts
var plugin = (def) => ({
  strict: false,
  alwaysApply: false,
  ...def,
  apply: (cldAsset, ctx) => def.apply(cldAsset, ctx, ctx)
});

// src/lib/utils.ts
var isArray = Array.isArray;
var entriesOf = Object.entries;

// src/plugins/underlays.ts
var UnderlaysPlugin = /* @__PURE__ */ plugin({
  name: "Underlays",
  supports: "all",
  inferOwnOptions: {},
  props: {
    underlay: true,
    underlays: true
  },
  apply: (cldAsset, opts) => {
    const { underlay, underlays = [] } = opts;
    const typeQualifier = "u";
    if (isArray(underlays)) {
      underlays.forEach(applyUnderlay);
    }
    if (typeof underlay === "string") {
      const underlayOptions = {
        publicId: underlay,
        crop: "fill",
        width: "1.0",
        height: "1.0",
        flags: ["relative"]
      };
      applyUnderlay(underlayOptions);
    }
    function applyUnderlay({
      publicId,
      type,
      position: position2,
      effects: layerEffects = [],
      flags: layerFlags = [],
      appliedFlags = [],
      ...options
    }) {
      const hasPublicId = typeof publicId === "string";
      const hasPosition = typeof position2 === "object";
      if (!hasPublicId) {
        console.warn(`An ${type} is missing a Public ID`);
        return;
      }
      let layerTransformation = `${typeQualifier}_${publicId.replace(
        /\//g,
        ":"
      )}`;
      const primary2 = [];
      const applied = [];
      entriesOf(options).forEach(([key, value]) => {
        if (!objectHasKey(primary, key)) return;
        const { qualifier } = primary[key];
        primary2.push(`${qualifier}_${value}`);
      });
      layerEffects.forEach((effect) => {
        Object.keys(effect).forEach((key) => {
          if (!objectHasKey(primary, key)) return;
          const { qualifier } = primary[key];
          primary2.push(`${qualifier}_${effect[key]}`);
        });
      });
      if (hasPosition) {
        entriesOf(position2).forEach(([key, value]) => {
          if (!objectHasKey(position, key)) return;
          const { qualifier } = position[key];
          applied.push(`${qualifier}_${value}`);
        });
      }
      const activeLayerFlags = isArray(layerFlags) ? layerFlags : [layerFlags];
      activeLayerFlags.forEach((flag) => primary2.push(`fl_${flag}`));
      const activeAppliedFlags = isArray(appliedFlags) ? appliedFlags : [appliedFlags];
      activeAppliedFlags.forEach((flag) => applied.push(`fl_${flag}`));
      layerTransformation = `${layerTransformation},${primary2.join(",")}`;
      layerTransformation = `${layerTransformation}/fl_layer_apply,fl_no_overflow`;
      if (applied.length > 0) {
        layerTransformation = `${layerTransformation},${applied.join(",")}`;
      }
      cldAsset.addTransformation(layerTransformation);
    }
    return {};
  }
});
export {
  UnderlaysPlugin
};
