// @cloudinary-util/url-loader/dist/plugins/version.cjs
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/plugins/version.ts
var version_exports = {};
__export(version_exports, {
  VersionPlugin: () => VersionPlugin
});
module.exports = __toCommonJS(version_exports);

// src/lib/plugin.ts
var plugin = (def) => ({
  strict: false,
  alwaysApply: false,
  ...def,
  apply: (cldAsset, ctx) => def.apply(cldAsset, ctx, ctx)
});

// src/plugins/version.ts
var VersionPlugin = /* @__PURE__ */ plugin({
  name: "Version",
  supports: "all",
  inferOwnOptions: {},
  props: {
    version: true
  },
  apply: (cldAsset, opts) => {
    const { version } = opts;
    if (typeof version !== "string" && typeof version !== "number") return {};
    cldAsset.setVersion(`${version}`.replace("v", ""));
    return {};
  }
});
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  VersionPlugin
});
