// @cloudinary-util/url-loader/dist/plugins/zoompan.cjs
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/plugins/zoompan.ts
var zoompan_exports = {};
__export(zoompan_exports, {
  ZoompanPlugin: () => ZoompanPlugin
});
module.exports = __toCommonJS(zoompan_exports);

// src/lib/plugin.ts
var plugin = (def) => ({
  strict: false,
  alwaysApply: false,
  ...def,
  apply: (cldAsset, ctx) => def.apply(cldAsset, ctx, ctx)
});

// src/plugins/zoompan.ts
var ZoompanPlugin = /* @__PURE__ */ plugin({
  name: "Zoompan",
  supports: "image",
  inferOwnOptions: {},
  props: {
    zoompan: true
  },
  apply: (cldAsset, opts) => {
    const { zoompan = false } = opts;
    const overrides = {
      format: void 0
    };
    if (zoompan === true) {
      cldAsset.addTransformation("e_zoompan");
    } else if (typeof zoompan === "string") {
      if (zoompan === "loop") {
        cldAsset.addTransformation("e_zoompan");
        cldAsset.addTransformation("e_loop");
      } else {
        cldAsset.addTransformation(`e_zoompan:${zoompan}`);
      }
    } else if (typeof zoompan === "object") {
      let zoompanEffect = "e_zoompan";
      if (typeof zoompan.options === "string") {
        zoompanEffect = `${zoompanEffect}:${zoompan.options}`;
      }
      cldAsset.addTransformation(zoompanEffect);
      let loopEffect;
      if (zoompan.loop === true) {
        loopEffect = "e_loop";
      } else if (typeof zoompan.loop === "string" || typeof zoompan.loop === "number") {
        loopEffect = `e_loop:${zoompan.loop}`;
      }
      if (loopEffect) {
        cldAsset.addTransformation(loopEffect);
      }
    }
    if (zoompan !== false) {
      overrides.format = "auto:animated";
    }
    return {
      options: overrides
    };
  }
});
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  ZoompanPlugin
});
