// @cloudinary-util/url-loader/dist/plugins/zoompan.js
// src/lib/plugin.ts
var plugin = (def) => ({
  strict: false,
  alwaysApply: false,
  ...def,
  apply: (cldAsset, ctx) => def.apply(cldAsset, ctx, ctx)
});

// src/plugins/zoompan.ts
var ZoompanPlugin = /* @__PURE__ */ plugin({
  name: "Zoompan",
  supports: "image",
  inferOwnOptions: {},
  props: {
    zoompan: true
  },
  apply: (cldAsset, opts) => {
    const { zoompan = false } = opts;
    const overrides = {
      format: void 0
    };
    if (zoompan === true) {
      cldAsset.addTransformation("e_zoompan");
    } else if (typeof zoompan === "string") {
      if (zoompan === "loop") {
        cldAsset.addTransformation("e_zoompan");
        cldAsset.addTransformation("e_loop");
      } else {
        cldAsset.addTransformation(`e_zoompan:${zoompan}`);
      }
    } else if (typeof zoompan === "object") {
      let zoompanEffect = "e_zoompan";
      if (typeof zoompan.options === "string") {
        zoompanEffect = `${zoompanEffect}:${zoompan.options}`;
      }
      cldAsset.addTransformation(zoompanEffect);
      let loopEffect;
      if (zoompan.loop === true) {
        loopEffect = "e_loop";
      } else if (typeof zoompan.loop === "string" || typeof zoompan.loop === "number") {
        loopEffect = `e_loop:${zoompan.loop}`;
      }
      if (loopEffect) {
        cldAsset.addTransformation(loopEffect);
      }
    }
    if (zoompan !== false) {
      overrides.format = "auto:animated";
    }
    return {
      options: overrides
    };
  }
});
export {
  ZoompanPlugin
};
