// astro-cloudinary/dist/chunk-HJHAQYNH.js
// src/constants/analytics.ts
import astroPkg from "astro/package.json";

// package.json
var package_default = {
  name: "astro-cloudinary",
  version: "1.3.5",
  type: "module",
  license: "MIT",
  module: "./index.ts",
  keywords: [
    "cloudinary",
    "astro",
    "astro-loader"
  ],
  exports: {
    ".": {
      import: "./index.ts"
    },
    "./helpers": {
      types: "./dist/helpers/index.d.ts",
      import: "./dist/helpers/index.js"
    },
    "./loaders": {
      types: "./dist/loaders/index.d.ts",
      import: "./dist/loaders/index.js"
    },
    "./package.json": "./package.json"
  },
  files: [
    "dist",
    "helpers",
    "loaders",
    "src",
    "index.ts"
  ],
  scripts: {
    dev: "tsup --watch",
    build: "tsup",
    prepublishOnly: "cp ../README.md . && cp ../LICENSE . && pnpm build",
    postpublish: "rm ./README.md && rm ./LICENSE"
  },
  dependencies: {
    "@cloudinary-util/types": "1.6.0",
    "@cloudinary-util/url-loader": "6.0.0",
    "@cloudinary-util/util": "^4.1.0",
    "@cloudinary/url-gen": "^1.20.0",
    "@unpic/astro": "^0.0.47",
    "@unpic/core": "^0.0.49",
    tsup: "^8.4.0",
    unpic: "^3.18.0"
  },
  peerDependencies: {
    astro: "^3.2.0 || ^4.0.0 || ^5.0.0"
  },
  devDependencies: {
    astro: "^5.11.0"
  }
};

// src/constants/analytics.ts
var ASTRO_CLOUDINARY_ANALYTICS_PRODUCT_ID = "B";
var ASTRO_CLOUDINARY_ANALYTICS_ID = "G";
var ASTRO_VERSION = normalizeVersion(astroPkg.version);
var ASTRO_CLOUDINARY_VERSION = normalizeVersion(package_default.version);
function normalizeVersion(version) {
  let normalized = version;
  if (normalized.includes("-")) {
    normalized = normalized.split("-")[0];
  }
  return normalized;
}

export {
  ASTRO_CLOUDINARY_ANALYTICS_PRODUCT_ID,
  ASTRO_CLOUDINARY_ANALYTICS_ID,
  ASTRO_VERSION,
  ASTRO_CLOUDINARY_VERSION
};
